src/rpc/xmlrpc_tinyxml2.cc

```diff
@@ -0,0 +1,451 @@
+// You must obey the GNU General Public License in all respects for
+// all of the code used other than OpenSSL.  If you modify file(s)
+// with this exception, you may extend this exception to your version
+// of the file(s), but you are not obligated to do so.  If you do not
+// wish to do so, delete this exception statement from your version.
+// If you delete this exception statement from all source files in the
+// program, then also delete it here.
+//
+// Contact:  Jari Sundell <jaris@ifi.uio.no>
+//
+//           Skomakerveien 33
+//           3185 Skoppum, NORWAY
+
+#include "config.h"
+
+#ifdef HAVE_XMLRPC_TINYXML2
+#ifdef HAVE_XMLRPC_C
+#error HAVE_XMLRPC_C and HAVE_XMLRPC_TINYXML2 cannot be used together. Please choose only one
+#endif
+
+#include <cctype>
+#include <string>
+#include <initializer_list>
+
+#include <stdlib.h>
+
+#include <rak/string_manip.h>
+#include <torrent/object.h>
+#include <torrent/exceptions.h>
+
+#include "rpc/tinyxml2/tinyxml2.h"
+#include "utils/base64.h"
+#include "xmlrpc.h"
+#include "parse_commands.h"
+
+namespace rpc {
+
+// Taken from xmlrpc-c
+const int XMLRPC_INTERNAL_ERROR = -500;
+const int XMLRPC_TYPE_ERROR = -501;
+const int XMLRPC_INDEX_ERROR = -502;
+const int XMLRPC_PARSE_ERROR = -503;
+const int XMLRPC_NETWORK_ERROR = -504;
+const int XMLRPC_TIMEOUT_ERROR = -505;
+const int XMLRPC_NO_SUCH_METHOD_ERROR = -506;
+const int XMLRPC_REQUEST_REFUSED_ERROR = -507;
+const int XMLRPC_INTROSPECTION_DISABLED_ERROR = -508;
+const int XMLRPC_LIMIT_EXCEEDED_ERROR = -509;
+const int XMLRPC_INVALID_UTF8_ERROR = -510;
+
+class xmlrpc_error : public torrent::base_error {
+public:
+  xmlrpc_error(int type, std::string msg) : m_type(type), m_msg(msg) {}
+  virtual ~xmlrpc_error() throw() {}
+
+  virtual int         type() const throw() { return m_type; }
+  virtual const char* what() const throw() { return m_msg.c_str(); }
+
+private:
+  int                 m_type;
+  std::string         m_msg;
+};
+
+const tinyxml2::XMLElement*
+element_access(const tinyxml2::XMLElement* elem, std::initializer_list<std::string> names) {
+  // Helper function to check each step of a element access, in lieu of XPath
+  const tinyxml2::XMLElement* result = elem;
+  for (auto itr : names) {
+    result = result->FirstChildElement(itr.c_str());
+    if (result == nullptr) {
+      throw xmlrpc_error(XMLRPC_PARSE_ERROR, "could not find expected element " + itr);
+    }
+  }
+  return result;
+}
+
+torrent::Object
+xml_value_to_object(const tinyxml2::XMLNode* elem) {
+  if (elem == nullptr) {
+    throw xmlrpc_error(XMLRPC_INTERNAL_ERROR, "received null element to convert");
+  }
+  if (std::strncmp(elem->Value(), "value", 5) != 0) {
+    throw xmlrpc_error(XMLRPC_INTERNAL_ERROR, "received non-value element to convert");
+  }
+  auto value_element = elem->FirstChild();
+  auto value_element_type = std::string(value_element->Value());
+  if (value_element_type == "string") {
+    auto value_element_child = value_element->FirstChild();
```

Comment:
I mainly wanted to avoid confusion with the doc->RootElement(), and was intended to reflect the `<value>X</value>` element that was passed in, will change though.

Reply:
Parent might be an alternative.

This is the difference between naming it what it is vs. what role it plays. E.g. calling someone a parent is often more descriptive than calling them a human.